import grpc
import sys
import os
import json
import time

# æ·»åŠ  proto ç”Ÿæˆä»£ç çš„è·¯å¾„
sys.path.append(os.path.join(os.path.dirname(__file__), 'proto_gen'))

try:
    import adapter_pb2
    import adapter_pb2_grpc
except ImportError:
    print("âŒ é”™è¯¯: æ‰¾ä¸åˆ°ç”Ÿæˆçš„ Proto ä»£ç ã€‚")
    sys.exit(1)

def run_test():
    print("ğŸ”Œ è¿æ¥ gRPC æœåŠ¡ (localhost:50051)...")
    channel = grpc.insecure_channel('localhost:50051')
    stub = adapter_pb2_grpc.AdapterServiceStub(channel)

    prompt_text = "è¯·ä½¿ç”¨ Bing æœç´¢ Rust è¯­è¨€ç›®å‰çš„æœ€æ–°ç¨³å®šç‰ˆæœ¬æ˜¯å¤šå°‘ã€‚åˆ›å»ºä¸€ä¸ªåä¸º rust_info.txt çš„æ–‡ä»¶ï¼ŒæŠŠç‰ˆæœ¬å·å†™è¿›å»ã€‚æœ€åï¼Œè¯·è¯»å–å¹¶æ˜¾ç¤ºè¯¥æ–‡ä»¶çš„å†…å®¹ï¼Œä»¥ç¡®è®¤å†™å…¥æˆåŠŸã€‚"
    
    req = adapter_pb2.RunRequest(
        request_id=f"test-{int(time.time())}",
        prompt=prompt_text,
        session_config=adapter_pb2.SessionConfig(
            model="qwen-plus",
            provider=adapter_pb2.ModelProviderInfo(
                name="aliyun",
                base_url="https://dashscope.aliyuncs.com/compatible-mode/v1",
                env_key="DASH_API_KEY", 
                wire_api=adapter_pb2.WireApi.WIRE_API_CHAT, 
                requires_openai_auth=False 
            ),
            sandbox_policy=adapter_pb2.SandboxPolicy.WORKSPACE_WRITE,
            approval_policy=adapter_pb2.ApprovalPolicy.NEVER,
            cwd="/tmp/codex_test_wd",
            mcp_servers={
                "bing-cn-mcp-server": adapter_pb2.McpServerDef(
                    server_type="streamable_http",
                    url="https://mcp.api-inference.modelscope.net/66eae62e82fe40/mcp"
                )
            }
        ),
        env_vars={"DASH_API_KEY": "sk-4438e8cfa0494e17b93845b7aa8b0bab"}
    )

    print(f"ğŸ“ ä»»åŠ¡: {prompt_text}\n" + "="*50)

    try:
        for response in stub.Run(req):
            if response.HasField("adapter_log"):
                # è¿‡æ»¤æ‰ SSE è¿™ç§å¿ƒè·³æ—¥å¿—ï¼Œåªçœ‹å…³é”®æ—¥å¿—
                log = response.adapter_log
                if "sse_event" not in log and "otel_manager" not in log:
                    print(f"ğŸ“‹ [SYSTEM] {log.strip()}")
            
            elif response.HasField("error"):
                print(f"âŒ [ERROR] {response.error}")
            
            elif response.HasField("codex_event_json"):
                try:
                    data = json.loads(response.codex_event_json)
                    msg = data.get("msg", {})
                    method = data.get("method")
                    
                    # é€‚é… Codex App-Server åè®®é£æ ¼
                    if method == "item/agentMessage/delta":
                        content = data.get("params", {}).get("delta", "")
                        if content:
                            sys.stdout.write(content)
                            sys.stdout.flush()
                    
                    elif method == "item/started":
                        item = data.get("params", {}).get("item", {})
                        item_type = item.get("type")
                        if item_type == "reasoning":
                            print(f"\nğŸ§  [æ€è€ƒ] {item.get('summary', 'æ€è€ƒä¸­...')}")
                        elif item_type == "commandExecution":
                            print(f"\nğŸ› ï¸  [æ‰§è¡Œå‘½ä»¤] {item.get('command')}")
                        elif item_type == "mcpToolCall":
                            print(f"\nğŸ”— [è°ƒç”¨å·¥å…·] {item.get('tool')}")
                    
                    elif method == "item/completed":
                        item = data.get("params", {}).get("item", {})
                        if item.get("type") == "commandExecution":
                            print(f"\nâœ… [æ‰§è¡Œç»“æœ] ExitCode: {item.get('exitCode')}")
                    
                    elif method == "turn/completed":
                        print("\n\nğŸ [ä»»åŠ¡å®Œæˆ]")
                    
                    # å¦‚æœæ˜¯åŸå§‹ JSONL æ ¼å¼ (Exec æ¨¡å¼)
                    elif "type" in data:
                        t = data["type"]
                        if t == "message" and "content" in data:
                            print(f"\nğŸ’¬ [Agent] {data['content']}")
                        elif t == "reasoning":
                            print(f"\nğŸ§  [æ€è€ƒ] {data.get('content')}")

                except Exception as e:
                    # å®åœ¨è§£æä¸åŠ¨ï¼Œä½†åˆä¸æ˜¯å™ªéŸ³ï¼Œå°±æ‰“å‡ºæ¥
                    # print(f"DEBUG: {response.codex_event_json}") 
                    pass

    except grpc.RpcError as e:
        print(f"\nâŒ gRPC è°ƒç”¨å¤±è´¥: {e.details()}")

    print("\n" + "="*50 + "\nâœ… æµ‹è¯•ç»“æŸ")

if __name__ == "__main__":
    run_test()
